query-algebrizer/src/clauses/not.rs

```diff
@@ -0,0 +1,500 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::btree_map::Entry;
+
+use mentat_core::Schema;
+
+use mentat_query::{NotJoin, UnifyVars, ContainsVariables};
+
+use clauses::ConjoiningClauses;
+
+use errors::{
+    Result,
+    ErrorKind,
+};
+
+use types::{ColumnConstraint, ComputedTable, EmptyBecause};
+
+impl ConjoiningClauses {
+    pub fn apply_not_join(&mut self, schema: &Schema, not_join: NotJoin) -> Result<()> {
+        let unified = match not_join.unify_vars {
+            UnifyVars::Implicit => not_join.collect_mentioned_variables(),
+            UnifyVars::Explicit(vs) => vs.into_iter().collect(),
+        };
+
+        let mut template = self.use_as_template(&unified);
+
+        for clause in not_join.clauses.into_iter() {
+            template.apply_clause(&schema, clause)?;
+        }
+
+        if template.is_known_empty() {
+            return Ok(());
+        } else {            
+            for v in unified.iter() {
```

Comment:
I just realized: you should do this _before_ you apply clauses.

There are two reasons for that.

Firstly, here you're providing static values! We want to use them when applying clauses. If you copy in `value_bindings` after applying patterns, it's too late! CC won't be able to use them to determine failure or inline values.

Secondly, you'll pull the value that we know — the external column association — towards the front of the mess of conditionals that we produce. That should produce a slightly more efficient query.

Doing that also turns the `is_known_empty` check into a simple gate, rather than having this `else`.